c_src/sasl_auth.c

```diff
@@ -713,19 +728,24 @@ static ERL_NIF_TERM sasl_kinit(ErlNifEnv* env, int UNUSED(argc), const ERL_NIF_T
         goto kinit_finish;
     }
 
-    if ((error = krb5_parse_name(context, (const char*)principal_mut, &principal)) != 0) {
+    if ((error = krb5_parse_name(context, (const char*)principal_char, &principal)) != 0) {
         tag = "krb5_parse_name";
         goto kinit_finish;
     }
 
-    if ((error = krb5_kt_resolve(context, (const char*)keytab_in, &kt_handle)) != 0) {
+    if ((error = krb5_kt_resolve(context, (const char*)keytab_char, &kt_handle)) != 0) {
         tag = "krb5_kt_resolve";
         goto kinit_finish;
     }
 
     handle_alive = 1;
 
-    if ((error = krb5_cc_default(context, &defcache)) != 0) {
+    /* NOTWORKING: (error = krb5_cc_resolve(context, (const char*)ccname_char, &ccache))
+     *
+     * krb5 doc says krb5_cc_default is essentially krb5_cc_resolve with default ccname, but it does not work.
+     * So we set environment variable KRB5CCNAME and call krb5_cc_default instead */
+    setenv("KRB5CCNAME", (const char*)ccname_char, 1);
```

Comment:
This in fact might be the simple reason why this didn't work for me when I experimented with it previously. Great find!

Reply:
Ok, better news, this does work, but you have to set default name first, there might be a better way, but : 

```
 if ((error = krb5_cc_set_default_name(context, (const char*)ccname_char)) != 0) {
        tag = "krb5_cc_set_default_name";
        goto kinit_finish;
    }

    if ((error = krb5_cc_default(context, &ccache)) != 0) {
        tag = "krb5_cc_default";
        goto kinit_finish;
    }
```

And voila, we don't have to rely on os envs now 😄 We'll need to look at caveats though around threads and such, though krb5 seems pretty thread safe.

Reply:
Better solution indeed, so you were half way there, but we need to make this call first : https://github.com/krb5/krb5/blob/b9b654e5b469140d5603f27af5bf83ee9a826349/src/lib/krb5/ccache/t_cc.c#L621C40-L621C51

Looks like there's a default memory cc struct (krb5_cc_ops) in an include somewhere.

Reply:
^ ignore, krb5_cc_resolve works fine, only need to register if it's a truly new cc type. 

```
   if ((error = krb5_cc_resolve(context, (const char*)ccname_char, &ccache)) != 0) {
        tag = "krb5_cc_set_default_name";
        goto kinit_finish;
    }

    printf("CC TYPE : %s\r\n", krb5_cc_get_type(context, ccache));
```
then 

```shell
1> example:main(client1).
CC TYPE : MEMORY
ok
2> 
```

Reply:
`krb5_cc_get_type` probably will return what you want to see,
but what matters is the return value of `sasl_client_new`.